### PR TITLE
 Added rewrite rule support instead of basepath match

### PR DESCRIPTION
**Pull request by mangchiandjjoe**

### CODE REVIEW

**Comment by mangchiandjjoe** [comment]
PTAL

It seems that std::regex works differently in gubuntu.
Matching /api/(.*) for /api/shelves are different.

in gubuntu:
0: /api/shelves
1: /shelves

in ubuntu:
0: /api/shelves
1: shelves

File: src/nginx/t/api_basepath.t [hunk 1]
```diff
@@ -63,7 +63,9 @@ service_control_config {
   }
 }
 api_service_config {
-  base_path: "/api"
+  rewrite: "/api/(.*) /\$1"
+  rewrite: "/apis/(.*) /\$1"
```
Comment (qiwzhang) [inline src/nginx/t/api_basepath.t]: second rule to do replacement, such as

 /foo/(.*)  /bar/$1

File: src/api_manager/rewrite_rule.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    for (auto it = replacement_parts_.begin(); it != replacement_parts_.end();
+         ++it) {
+      switch (it->type) {
+        case ReplacementPartType::TEXT:
+          ss << it->text;
+          break;
+        case ReplacementPartType::REPLACEMENT:
+          if (it->index >= 0 && it->index < (int)cm.size()) {
+            ss << cm[it->index];
+          }
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.cc]: That case should not be happen. I added default section with error log.

File: src/api_manager/proto/server_config.proto [hunk 3]
```diff
@@ -210,15 +210,17 @@ message ServiceConfigRollout {
...
+  // sequentially in order of their appearance in the configuration file.
+  // Syntax:  [regex] replacement
+  // regex follows ECMAScript regular expressions pattern syntax
+  repeated string rewrite = 1;
+
+  // Enables or disables logging of ESP rewrite processing results into the
+  // error_log at the info level.
   // The default value is false
-  bool base_path_hard_match = 2;
+  bool rewrite_log = 2;
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/proto/server_config.proto]: Removed configuration and added debug header

File: src/api_manager/api_manager_impl.h [hunk 4]
```diff
@@ -103,6 +106,8 @@ class ApiManagerImpl : public ApiManager {
   // A config manager will be initialized when server_config.rollout_strategy is
   // set to "managed"
   std::unique_ptr<ConfigManager> config_manager_;
+
+  std::vector<RewriteRule *> rewrite_rules_;
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/api_manager_impl.h]: Done

File: src/api_manager/api_manager_impl.h [hunk 5]
```diff
@@ -72,8 +75,8 @@ class ApiManagerImpl : public ApiManager {
   // REJECT - Return 404
   // REWRITE - Update url and unparsed_uri with destination_url
   // NONE - Do nothing
-  ApiManager::ApiBasepathRewriteAction ReWriteURL(
-      const std::string &url, std::string *destination_url) override;
+  ApiManager::RewriteAction ReWriteURL(const std::string &url,
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/api_manager_impl.h]: Done

File: src/api_manager/api_manager_impl.cc [hunk 6]
```diff
@@ -36,6 +36,35 @@ ApiManagerImpl::ApiManagerImpl(std::unique_ptr<ApiManagerEnvInterface> env,
...
+                                               global_context_->env(),
+                                               global_context_->server_config()
+                                                   ->api_service_config()
+                                                   .rewrite_log()));
+    }
+  }
+}
+
+ApiManagerImpl::~ApiManagerImpl() {
+  for (auto rule : rewrite_rules_) {
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/api_manager_impl.cc]: Switched to std::vector<std::unique_ptr<RewriteRule>>

File: src/api_manager/api_manager_impl.cc [hunk 7]
```diff
@@ -36,6 +36,35 @@ ApiManagerImpl::ApiManagerImpl(std::unique_ptr<ApiManagerEnvInterface> env,
...
   check_workflow_->RegisterAll();
+
+  if (global_context_->server_config() &&
+      global_context_->server_config()->has_api_service_config()) {
+    for (auto rule :
+         global_context_->server_config()->api_service_config().rewrite()) {
+      std::stringstream ss(rule);
+      std::istream_iterator<std::string> begin(ss);
+      std::istream_iterator<std::string> end;
+      std::vector<std::string> parts(begin, end);
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/api_manager_impl.cc]: My plan is to add an array argument "rewrite".
--rewrite "/api/(.*) /$1" --rewrite "/apis/(.*) /$1"

[the remark this answers is not in the document]
Comment (qiwzhang) [reply src/api_manager/api_manager_impl.cc]: I still don't like space .    How about bar "|"  or ":",  check with URL path  RFC, if they are not legal in URL path,  we can use one of them as our delimiter.

[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/api_manager_impl.cc]: Both nginx and apache rewrite rule modules use white space as an delimiter.
I think white space is good here.
':' is resolved for the port number. we might be able to use the '|' for the delimiter.

File: src/api_manager/rewrite_rule.cc [hunk 8]
```diff
@@ -90,77 +135,153 @@ RewriteRule::RewriteRule(std::string regex, std::string replacement,
...
+
+  std::vector<std::string> match;
+
+  const char *psub_str_match_str;
+  for (int j = 0; j < pcre_exec_ret; j++) {
+    pcre_get_substring(uri.c_str(), sub_str_vec, pcre_exec_ret, j,
+                       &(psub_str_match_str));
+    match.push_back(psub_str_match_str);
+  }
+  pcre_free_substring(psub_str_match_str);
```
Comment (qiwzhang) [inline src/api_manager/rewrite_rule.cc]: instead of making a copy,  can we do this in line 266

File: src/nginx/config.cc [hunk 9]
```diff
@@ -593,7 +602,15 @@ ngx_int_t ngx_esp_build_server_config(ngx_conf_t *cf, ngx_esp_loc_conf_t *lc,
 
   // validate rewrite rule syntax
   if (config.has_api_service_config()) {
-    std::unique_ptr<std::regex> rule_check;
+    pcre *regex_compiled_;
+    pcre_extra *regex_extra_;
+
+    // backup functional pointers of pcre memory functions
+    origin_pcre_malloc = pcre_malloc;
```
Comment (qiwzhang) [inline src/nginx/config.cc]: please try to re-use the code in rewrite.cc

File: src/nginx/config.cc [hunk 10]
```diff
@@ -593,7 +602,15 @@ ngx_int_t ngx_esp_build_server_config(ngx_conf_t *cf, ngx_esp_loc_conf_t *lc,
 
   // validate rewrite rule syntax
   if (config.has_api_service_config()) {
-    std::unique_ptr<std::regex> rule_check;
+    pcre *regex_compiled_;
```
Comment (qiwzhang) [inline src/nginx/config.cc]: this function is getting too big,  please create a new function to handle rewrite()

File: src/nginx/config.cc (hunk 9 above)
Comment (qiwzhang) [reply src/nginx/config.cc]: may be move these verification code into rewrite.cc?

File: src/api_manager/proto/server_config.proto [hunk 11]
```diff
@@ -210,17 +210,22 @@ message ServiceConfigRollout {
...
+  // A rewrite consists of two arguments separated by spaces. The arguments are:
+  //
+  //  pattern: which incoming URLs should be affected by the rule. PCRE
+  //    compatible regular expression checked against the request
+  //    URI, which is the part after http://hostname
+  //  replacement: where should the matching requests be sent. Modification to
+  //    the part matched by the pattern or absolute path. it doesn't allow
+  //    destinations starting with either "http://" or "https://"
+  //
+  // ex. "^/?([a-z]+)/(.*)$ /api.foo?page=$1&id=$2"
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/proto/server_config.proto]: Updated to "Replacement starts with either http:// or https:// is not allowed"

File: src/api_manager/rewrite_rule.cc [hunk 12]
```diff
@@ -90,77 +135,153 @@ RewriteRule::RewriteRule(std::string regex, std::string replacement,
...
+    restore_regex_malloc_free();
+    return false;
+  }
+
+  if (debug_mode) {
+    rewrite_log << kEspRewriteTitle << ": matching rule: " << regex_pattern_
+                << " " << this->replacement_ << std::endl;
+  }
+
+  // At this point, rc contains the number of substring matches found...
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.cc]: Update comment to "pcre_exec_ret"

File: src/api_manager/rewrite_rule.cc (hunk 8 above)
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.cc]: Applied

File: src/nginx/config.cc (hunk 9 above)
Comment (mangchiandjjoe) [reply src/nginx/config.cc]: Created a static method, RewriteRule::ValidateRewriteRule

**Comment by mangchiandjjoe** [comment]
PTAL

File: src/api_manager/rewrite_rule.cc [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+                   std::string(pcre_error_str));
+
+    pcre_free(regex_compiled_);
+    return;
+  }
+
+  restore_regex_malloc_free();
+
+  std::string segment;
+  ReplacementPartType status = ReplacementPartType::TEXT;
```
Comment (qiwzhang) [inline src/api_manager/rewrite_rule.cc]: Like to see unit-test for this portion of code

File: src/nginx/module.cc [hunk 14]
```diff
@@ -42,45 +42,70 @@
...
+  }
 
-  if (unparsed_uri_action == ApiManager::ApiBasepathRewriteAction::REWRITE) {
+  // set rewrite debug mode
+  std::transform(debug_header.begin(), debug_header.end(), debug_header.begin(),
+                 ::tolower);
+  bool debug_mode = debug_header == "true";
+
+  std::string unparsed_uri;
+  if (lc->esp->ReWriteURL(ngx_str_to_std(r->unparsed_uri), &unparsed_uri,
```
Comment (qiwzhang) [inline src/nginx/module.cc]: can we not to make a copy of unparsed_uri?   some requests could be very long.  try to see if we can use

::google::protobuf::StringPiece

or only make a copy when rewrite is needed.

File: src/nginx/module.cc [hunk 15]
```diff
@@ -42,45 +42,70 @@
...
-ApiManager::ApiBasepathRewriteAction ngx_esp_rewrite_uri(
-    ngx_http_request_t *r, ngx_esp_loc_conf_t *lc) {
-  std::string unparsed_uri;
-  ApiManager::ApiBasepathRewriteAction unparsed_uri_action =
-      lc->esp->ReWriteURL(ngx_str_to_std(r->unparsed_uri), &unparsed_uri);
+// Internal debugging header
+std::string kXEndpointsDebugUrlRewrite = "x-endpoints-debug-url-rewrite";
+
+// decode encoded uri
+void url_decode(const std::string &uri, std::string &decoded) {
```
Comment (qiwzhang) [inline src/nginx/module.cc]: did you write this on your own?  can we copy Nginx code here?

File: src/nginx/module.cc [hunk 16]
```diff
@@ -42,45 +42,70 @@
...
+  decoded.reserve(uri.length());
+  for (std::size_t i = 0; i < uri.length(); i++) {
+    if (int(uri[i]) == 37) {
+      int ii;
+      sscanf(uri.substr(i + 1, 2).c_str(), "%x", &ii);
```
Comment (qiwzhang) [inline src/nginx/module.cc]: will this crash if % is the last one?

File: src/nginx/module.cc [hunk 17]
```diff
@@ -42,45 +42,70 @@
...
+      char ch = static_cast<char>(ii);
+      decoded.append(std::string(1, ch));
```
Comment (qiwzhang) [inline src/nginx/module.cc]: does std::string have append for a char, ?

File: src/nginx/module.cc [hunk 18]
```diff
@@ -42,45 +42,70 @@
...
-// Then returns ApiManager::ApiBasepathRewriteAction.
-// If the return value is ApiManager::ApiBasepathRewriteAction::REJECT,
-// then nginx should response status 404 NotFound
-ApiManager::ApiBasepathRewriteAction ngx_esp_rewrite_uri(
-    ngx_http_request_t *r, ngx_esp_loc_conf_t *lc) {
-  std::string unparsed_uri;
-  ApiManager::ApiBasepathRewriteAction unparsed_uri_action =
-      lc->esp->ReWriteURL(ngx_str_to_std(r->unparsed_uri), &unparsed_uri);
+// Internal debugging header
+std::string kXEndpointsDebugUrlRewrite = "x-endpoints-debug-url-rewrite";
```
Comment (qiwzhang) [inline src/nginx/module.cc]: use ngx_str_t may be easier to use than std::string

File: src/nginx/module.cc [hunk 19]
```diff
@@ -42,45 +42,70 @@
...
+      r, reinterpret_cast<u_char *>(
+             const_cast<char *>(kXEndpointsDebugUrlRewrite.c_str())),
+      kXEndpointsDebugUrlRewrite.length());
+  if (h && h->value.len > 0) {
+    debug_header.assign(ngx_str_to_std(h->value));
+  }
 
-  if (unparsed_uri_action == ApiManager::ApiBasepathRewriteAction::REWRITE) {
+  // set rewrite debug mode
+  std::transform(debug_header.begin(), debug_header.end(), debug_header.begin(),
```
Comment (qiwzhang) [inline src/nginx/module.cc]: not to call line 86-88 if the debug header doesn't present.  
This function is called for every request,  speed is very important.

File: src/nginx/module.cc (hunk 15 above)
Comment (mangchiandjjoe) [reply src/nginx/module.cc]: I added my own. 

I have searched the function from nginx code base but it seems that nginx doesn't have a simple one. 
ngx_http_parse_request_line and ngx_http_parse_complex_uri are decoding the request but it is complicated.

File: src/nginx/module.cc (hunk 16 above)
Comment (mangchiandjjoe) [reply src/nginx/module.cc]: Added more error handling.

File: src/nginx/module.cc (hunk 17 above)
Comment (mangchiandjjoe) [reply src/nginx/module.cc]: Replaced with decoded.append(1, ch)

File: src/nginx/module.cc (hunk 19 above)
Comment (mangchiandjjoe) [reply src/nginx/module.cc]: moved the transform code to the inside of if statement

File: src/nginx/module.cc (hunk 18 above)
Comment (mangchiandjjoe) [reply src/nginx/module.cc]: Done

File: src/api_manager/rewrite_rule.h [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    return Check(uri, destination, false);
+  }
+
+  bool Check(const std::string &uri, std::string *destination, bool debug_mode);
+
+  static bool ValidateRewriteRule(const std::string &rule,
+                                  std::string *error_msg);
+
+ private:
+  std::string regex_pattern_;
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.h]: Added comments.
Those are for debug logging, we need to keep it.

File: src/api_manager/rewrite_rule.h [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+  struct ReplacementSegment {
+    ReplacementPartType type;
+    int index;
+    std::string text;
+
+    ReplacementSegment(ReplacementPartType type, int index, std::string text)
+        : type(type), index(index), text(text) {}
+  };
+
+  RewriteRule(std::string regex, std::string replacement,
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.h]: Relocated.

File: src/api_manager/rewrite_rule.h [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+#include "src/api_manager/weighted_selector.h"
+
+namespace google {
+namespace api_manager {
+
+class RewriteRule {
+ public:
+  enum ReplacementPartType { TEXT, REPLACEMENT };
+
+  struct ReplacementSegment {
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.h]: Changed to private

File: src/api_manager/rewrite_rule.cc (hunk 13 above)
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.cc]: MatchAndReplacementPattern is testing this part.

Added more cases.

File: src/api_manager/rewrite_rule.cc [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+void restore_regex_malloc_free() {
+  pcre_malloc = origin_pcre_malloc;
+  pcre_free = origin_pcre_free;
+}
+
+}  // namespace
+
+bool RewriteRule::ValidateRewriteRule(const std::string &rule,
+                                      std::string *error_msg) {
+  backup_regex_malloc_free();
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.cc]: Create a class for scoped overriding.

File: src/api_manager/rewrite_rule.cc [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+
+namespace google {
+namespace api_manager {
+
+namespace {
+
+// module title
+const std::string kEspRewriteTitle = "esp_rewrite";
+
+const int kMaxRegexMathCount = 100;
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply src/api_manager/rewrite_rule.cc]: Done

File: src/nginx/module.cc (hunk 14 above)
Comment (mangchiandjjoe) [reply src/nginx/module.cc]: Removed std::string copy

**Comment by mangchiandjjoe** [comment]
PTAL

**Pull request merged**